src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,162 @@
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
+                `State '${state.name}' is not allowed on this StatefulActionButton.\nEnsure to use the object references used at the instantiation`
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initalState) {
+            console.warn('The stateful action button cannot render if the inital states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initalState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        return !this.options.allowedTransitions || this.areTransitionsValid();
+    }
+
+    private areTransitionsValid(): boolean {
+        for (let index = 0; index < this.options.allowedTransitions.length; index++) {
+            const transition = this.options.allowedTransitions[index];
+            if (this.options.states.indexOf(transition.from) == -1) {
+                console.warn(this.generateInvalidTransitionMessage(index, true));
+                return false;
+            }
+            if (this.options.states.indexOf(transition.from) == -1) {
+                console.warn(this.generateInvalidTransitionMessage(index, false));
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private generateInvalidTransitionMessage(transitionNumber: number, isOrigin: boolean) {
+        return `The stateful action button cannot render if its origin is not in the list of states:\n\t${
```

Comment:
Can you store this magic template in a const or a static ?

Reply:
It's in a function that does just that, no, really, **no**.